### PR TITLE
Implements the first version of the conversion

### PR DESCRIPTION
**Pull request by simonedavico**
Implements the first version of the conversion from a BenchFlow configuration file in YAML format to a Faban configuration file in XML format. A couple of notes:
- As of this version, error handling is not the most robust. It will throw exceptions in some cases in which it should fail, but they are generic exceptions which will be replaced by more explicative ones in the future;
- Is it possible that not all of the Faban XML nodes are correctly namespaced (all the ones I found on the docs page and in the examples we have are correctly namespaced).

###### How to use the converter ######
Class `BenchflowConfigConverter` implements a `from` method which takes the `InputStream` of the configuration file to be converted and returns a `scala.xml.Elem`, which is the generated XML tree.
An example usage can be found in the test folder.

The constructor of `BenchflowConfigConverter` currently takes two strings:
1. JVM location on the machine
2. JVM options separated by a space

In the future those two will be abstracted to a more general `ConverterConfiguration` which includes all the parameters we need, but for now it is enough.

This commit implements #3.

### CODE REVIEW

**Comment by VincenzoFerme** [comment]
@simonedavico in order to close the mentioned issue (#3) you still need to update the following file https://gist.github.com/VincenzoFerme/14db45b824dc34275874.

**Comment by VincenzoFerme** [comment]
@simonedavico transform the issues described in the pull request description, in actual issues on the repository.

File: benchflow-to-faban-config-converter/pom.xml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                        <goals>
+                            <goal>compile</goal>
+                            <goal>testCompile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <scalaVersion>${scala.version}</scalaVersion>
+                    <args>
+                        <arg>-target:jvm-1.5</arg>
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/pom.xml]: Why JVM 1.5?

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+          </fa:runConfig>
+        </xml>.copy(label = elem.label)
+      case _ => throw new Exception //TODO: throw meaningful exception
+    }
+  }
+
+}
+
+//the interface to the business logic
+class BenchflowConfigConverter(val javaHome: String, val javaOpts: String) {
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Lets add a configuration file (the one that DropWizard will use) with the following values and get rid of the parameters:

```yml
java_home: /usr/lib/jvm/java7
java_opts: -Xmx256m -Xms64m -XX:+DisableExplicitGC
```

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      case <drivers>{drivers @ _*}</drivers> =>
+        <xml>
+          <jvmConfig xmlns="http://faban.sunsource.net/ns/fabanharness">
+            <javaHome>{javaHome}</javaHome>
+            <jvmOptions>{javaOpts}</jvmOptions>
+          </jvmConfig>
+          <fa:runConfig definition={elem.label}
+                        xmlns:fa="http://faban.sunsource.net/ns/faban"
+                        xmlns:fh="http://faban.sunsource.net/ns/fabanharness"
+                        xmlns="http://faban.sunsource.net/ns/fabandriver">
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Get rid of the default namespace for the driver. Use: fd

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    elem.asInstanceOf[scala.xml.Elem].copy(label = ns + ":" + elem.label)
+  }
+
+  private def transformDriverConfig(elem: scala.xml.Node): scala.xml.Elem =
+    <driverConfig name={elem.label}>{elem.child map addFabanNamespace}</driverConfig>
+
+  def apply(elem: scala.xml.Elem, javaHome: String, javaOpts: String): scala.xml.Elem = {
+    (elem \ "drivers" theSeq).head match {
+      case <drivers>{drivers @ _*}</drivers> =>
+        <xml>
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: The Faban configuration xml has `<?xml version="1.0" encoding="UTF-8"?>`. Why you have a different one?

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    map map toXML toList
+  }
+
+}
+
+//transform intermediate XML into Faban compliant XML
+object FabanXMLTransformer {
+
+  def propertyToNamespace =
+    Map(
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: You miss: `fh:cpus`, `fh:enabled`, the `threadstart` section, the `stats` section

Reference: http://faban.org/1.3/docs/guide/harnessdev/configfile.html

If you think they are not needed in the first prototype, please open an issue reporting the fields that are not yet converted. Also other properties you found on the documentation/online forums

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  def toXML(map: Map[String, Any]): List[scala.xml.Elem] = {
+    map map toXML toList
+  }
+
+}
+
+//transform intermediate XML into Faban compliant XML
+object FabanXMLTransformer {
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: How can I set the [`serverconfig`](http://faban.org/1.3/docs/guide/harnessdev/configfile.html)

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala [hunk 7]
```diff
@@ -0,0 +1,65 @@
+package cloud.benchflow.config
+
+import scala.util.parsing.combinator._
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala]: Document why you did use an already implemented solution

Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala]: Document why you did use an already implemented solution

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 6 above)
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Decide, add in a configuration file, and document the properties that we set using default values. One of the properties to set by default is [`fh:timeSync`](http://faban.org/1.3/docs/guide/harnessdev/configfile.html). It should be set by default to false, because it is not possible inside Docker containers without getting sudo access to the host operating system.

File: benchflow-to-faban-config-converter/src/test/resources/test.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+goBenchmark:
```
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/test/resources/test.yaml]: Write a test that generates the following file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<wfmsbenchmark>
    <jvmConfig xmlns="http://faban.sunsource.net/ns/fabanharness">
        <javaHome>/usr/lib/jvm/java7</javaHome>
        <jvmOptions>-Xmx256m -Xms64m -XX:+DisableExplicitGC</jvmOptions>
    </jvmConfig>
    <fa:runConfig definition="cloud.benchflow.drivers.WfMSDriver"
                  xmlns:fa="http://faban.sunsource.net/ns/faban"
                  xmlns:fh="http://faban.sunsource.net/ns/fabanharness"
                  xmlns:fd="http://faban.sunsource.net/ns/fabandriver">
        <fh:description>A WfMS benchmark</fh:description>
        <fa:hostConfig>
            <fa:host>192.181.131.121</fa:host>
        </fa:hostConfig>
        <fa:scale>10</fa:scale>
        <fa:runControl unit="time">
            <fa:rampUp>30</fa:rampUp>
            <fa:steadyState>120</fa:steadyState>
            <fa:rampDown>30</fa:rampDown>
        </fa:runControl>
        <fd:driverConfig name="WfMSDriver">
            <fd:agents>192.181.131.121:10</fd:agents>
            <fd:properties>
                <fd:property>
                    <fd:name>contextPath</fd:name>
                    <fd:value>/</fd:value>
                </fd:property>
                <fd:property>
                    <fd:name>secure</fd:name>
                    <fd:value>false</fd:value>
                </fd:property>
            </fd:properties>
        </fd:driverConfig>
    </fa:runConfig>
    <sutconfig>
        <fa:hostConfig xmlns:fh="http://faban.sunsource.net/ns/fabanharness" 
                       xmlns:fa="http://faban.sunsource.net/ns/faban">
            <fa:hostPorts>192.181.131.120:6060</fa:hostPorts>
        </fa:hostConfig>
    </sutconfig>
</wfmsbenchmark>
```

File: benchflow-to-faban-config-converter/pom.xml (hunk 1 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/pom.xml]: I copy pasted from an old example. Corrected.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 2 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: I don't see how adding these informations in a configuration file would prevent passing them to the constructor of the class. I would still need to pass them.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 3 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Implemented.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 4 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: It slipped. Added.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 5 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Added namespaces for `cpus` and `enabled`. `threadStart` and `stats`as far as I can see don't have a namespace, so they are already ok.

I implemented all the fields I encountered so far, as I find new ones I will update the `propertyToNamespace` map accordingly.

Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Added also `fh:timeSync`.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 6 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: I think this shouldn't be part of this pull request. The requirement for this step was to implement a 1:1 mapping from YAML to XML, so there is still no concept of "default values" for properties.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala (hunk 7 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala]: This implementation is basically a conversion of the one discussed in chapter 33.4 of the book `Programming in Scala - Second Edition`, which fully implements a JSON parser using parser combinators. I implemented my own version, but it was lacking in some aspects (it didn't account for some cases since I didn't knew some possibilities offered by the YAML specification), so when I found this one I decided to swap it. It's very similar, but accounts for all cases (mine for example didn't account for the case in which the YAML file starts with list of properties). Changing mine would have required some changes that would have turned it in the exact copy of this one, so in the end it's the same.

I checked the implementation and it's referenced in several places online and includes some test cases, so I think it will work fine.

File: benchflow-to-faban-config-converter/src/test/resources/test.yaml (hunk 8 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/test/resources/test.yaml]: I added a test that should generate this configuration.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala (hunk 7 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/YAMLParser.scala]: Nevermind, the implementation had a serious bug on nested lists handling. I switched to the SnakeYAML implementation.

File: benchflow-to-faban-config-converter/src/test/resources/test.yaml (hunk 8 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/test/resources/test.yaml]: The `<?xml version="1.0" encoding="UTF-8"?>` declaration will be added once the generated XML is saved to a file, so it's not part of what is printed by the `QuickTest` object.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 2 above)
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Right

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 4 above)
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Test if Faban accepts the configuration file. If yes, then we can avoid the declaration.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 6 above)
Comment (VincenzoFerme) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: Open an issue about it then, please.

File: benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala (hunk 2 above)
Comment (simonedavico) [inline benchflow-to-faban-config-converter/src/main/scala/cloud/benchflow/config/BenchflowConfigConverter.scala]: I added a YAML config file with these informations which is read by the constructor of the class, so we don't need constructor parameters anymore.
